### PR TITLE
[Sol -> Yul] Fix struct member access for memory and implement for calldata.

### LINKED ISSUE #9174: [Sol->Yul] Codegen for read from struct member throws ICE
## Description

While writing the following test for #8858 (to test whether dirty bits in struct members are cleared before copy), I ran into the error below.

```
pragma experimental ABIEncoderV2;
contract C {
    struct S {
        uint8[] m;
    }
    function f(S memory s) public pure returns (bool correct) {
        assembly {
            mstore(add(s, 64), 257)
        }
        uint8 x = s.m[0];
        uint r;
        assembly {
            r := x
        }
        correct = r == 0x01;
    }
}
// ====
// compileViaYul: true
// ----
// f((uint8[])): 0x20, 0x20, 0x01, 0x00 -> true
```

throws 

https://github.com/ethereum/solidity/blob/2410f0ed5a47cf262a2d836f4c493c48e8bd9977/libsolidity/codegen/ir/IRVariable.cpp#L52

I believe the problem is in the assignment `uint8 x = s.m[0]` that reads from struct because the following test passes

```
pragma experimental ABIEncoderV2;
contract C {
    struct S {
        uint8[] m;
    }
    function f(S memory s) public pure returns (bool correct) {
        assembly {
            mstore(add(s, 64), 257)
        }
        //uint8 x = s.m[0];
        uint r;
        assembly {
            r := mload(add(s, 64))
        }
        correct = r == 0x0101;
    }
}
// ====
// compileViaYul: true
// ----
// f((uint8[])): 0x20, 0x20, 0x01, 0x00 -> true
```

## Environment

- Compiler version: latest develop

## Steps to Reproduce

```
$ solc --ir <input.sol>
```

where `input.sol` is the first code snippet.

### PR DESCRIPTION
**Pull request by ekpyron**
Fixes https://github.com/ethereum/solidity/issues/9174

### CODE REVIEW

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 1]
```diff
@@ -1497,7 +1497,25 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 		}
 		case DataLocation::CallData:
 		{
-			solUnimplementedAssert(false, "");
+			string baseRef = expression.part("offset").name();
+			string offset = m_context.newYulVariable();
+			m_code << "let " << offset << " := " << "add(" << baseRef << ", " << to_string(structType.calldataOffsetOfMember(member)) << ")\n";
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1502]: Doesn't this apply the offset twice if combined with `accessCalldataTailFunction`?

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1502]: The signature for ``accessCalldataTailFunction`` is ``function <functionName>(base_ref, ptr_to_tail) -> ...``... whyever that is... must have been convenient to do it like that somewhere else...

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1502]: Or is this correct? Let me doublecheck...

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1502]: In the test the struct lives at offset ``0x20``...
That's ``baseRef``...
``offset`` will become ``add(baseRef, 0x20)``, which is ``0x40`` - the offset of the tail pointer.
The value at 0x40 is the tail pointer - in the test that's ``0x60``... that's the first thing ``accessCalldataTailFunction`` reads... and then it adds ``baseRef`` to it to get the data pointer.
I think this is all correct.

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1502]: Maybe even another test with two calldata arguments makes sense, just to be sure.

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1502]: Always a mess with the calldata stuff, though... I wonder whether we could have done something about it differently and have avoided all that...

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1502]: [note Ah damn... ``add(baseRef, 0x20)`` is of course ``0x40`` if baseref is 0x20, not ``0x60``... I'm editing that comment above...]

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1502]: Added test with two calldata arguments - works as well.

**Review by mijovic — APPROVED** [review]
Looks good

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1502]: Thanks, looks good after looking it for a second time.

**Pull request merged**